# Experiments

In this notebook, we experiment and evaluate different configurations,
in order to find one which gives the best performance.
This includes tuning hyperparamaters, and trying various data augmentation strategies.

Across all configurations, we are attempting to finetune a RoBERTa model.

In [1]:
!pip install --quiet datasets
from modeleval import *

## Baseline
We first fine-tune RoBERTa on with no data augmentation strategy, for $3$ epochs, at a learning rate of $3 \times 10 ^{-5}$. We use a linear learning rate scheduler, and a batch size of $8$.

In [2]:
evaluate_model()

/usr/local/lib/python3.11/dist-packages/huggingface_hub/utils/_auth.py:94: UserWarning: 
The secret `HF_TOKEN` does not exist in your Colab secrets.
To authenticate with the Hugging Face Hub, create a token in your settings tab (https://huggingface.co/settings/tokens), set it as secret in your Google Colab and restart your session.
You will be able to reuse this secret in all of your notebooks.
Please note that authentication is recommended but still optional to access public models or datasets.
  warnings.warn(
Some weights of RobertaForSequenceClassification were not initialized from the model checkpoint at roberta-base and are newly initialized: ['classifier.dense.bias', 'classifier.dense.weight', 'classifier.out_proj.bias', 'classifier.out_proj.weight']
You should probably TRAIN this model on a down-stream task to be able to use it for predictions and inference.


Map:   0%|          | 0/8375 [00:00<?, ? examples/s]

Training on CUDA
Training fold 1...


Some weights of RobertaForSequenceClassification were not initialized from the model checkpoint at roberta-base and are newly initialized: ['classifier.dense.bias', 'classifier.dense.weight', 'classifier.out_proj.bias', 'classifier.out_proj.weight']
You should probably TRAIN this model on a down-stream task to be able to use it for predictions and inference.
/usr/local/lib/python3.11/dist-packages/transformers/training_args.py:1575: FutureWarning: `evaluation_strategy` is deprecated and will be removed in version 4.46 of 🤗 Transformers. Use `eval_strategy` instead
  warnings.warn(
/content/modeleval.py:78: FutureWarning: `tokenizer` is deprecated and will be removed in version 5.0.0 for `Trainer.__init__`. Use `processing_class` instead.
  trainer = Trainer(


Fold 1 evaluation results: {'eval_f1': 0.547945205479452, 'eval_loss': 0.35339587926864624, 'eval_runtime': 11.9286, 'eval_samples_per_second': 140.419, 'eval_steps_per_second': 17.605, 'epoch': 3.0}
Training fold 2...


Some weights of RobertaForSequenceClassification were not initialized from the model checkpoint at roberta-base and are newly initialized: ['classifier.dense.bias', 'classifier.dense.weight', 'classifier.out_proj.bias', 'classifier.out_proj.weight']
You should probably TRAIN this model on a down-stream task to be able to use it for predictions and inference.
/usr/local/lib/python3.11/dist-packages/transformers/training_args.py:1575: FutureWarning: `evaluation_strategy` is deprecated and will be removed in version 4.46 of 🤗 Transformers. Use `eval_strategy` instead
  warnings.warn(
/content/modeleval.py:78: FutureWarning: `tokenizer` is deprecated and will be removed in version 5.0.0 for `Trainer.__init__`. Use `processing_class` instead.
  trainer = Trainer(


Fold 2 evaluation results: {'eval_f1': 0.5531914893617021, 'eval_loss': 0.29567497968673706, 'eval_runtime': 11.7222, 'eval_samples_per_second': 142.891, 'eval_steps_per_second': 17.915, 'epoch': 3.0}
Training fold 3...


Some weights of RobertaForSequenceClassification were not initialized from the model checkpoint at roberta-base and are newly initialized: ['classifier.dense.bias', 'classifier.dense.weight', 'classifier.out_proj.bias', 'classifier.out_proj.weight']
You should probably TRAIN this model on a down-stream task to be able to use it for predictions and inference.
/usr/local/lib/python3.11/dist-packages/transformers/training_args.py:1575: FutureWarning: `evaluation_strategy` is deprecated and will be removed in version 4.46 of 🤗 Transformers. Use `eval_strategy` instead
  warnings.warn(
/content/modeleval.py:78: FutureWarning: `tokenizer` is deprecated and will be removed in version 5.0.0 for `Trainer.__init__`. Use `processing_class` instead.
  trainer = Trainer(


Fold 3 evaluation results: {'eval_f1': 0.535593220338983, 'eval_loss': 0.3302139639854431, 'eval_runtime': 11.7422, 'eval_samples_per_second': 142.648, 'eval_steps_per_second': 17.884, 'epoch': 3.0}
Training fold 4...


Some weights of RobertaForSequenceClassification were not initialized from the model checkpoint at roberta-base and are newly initialized: ['classifier.dense.bias', 'classifier.dense.weight', 'classifier.out_proj.bias', 'classifier.out_proj.weight']
You should probably TRAIN this model on a down-stream task to be able to use it for predictions and inference.
/usr/local/lib/python3.11/dist-packages/transformers/training_args.py:1575: FutureWarning: `evaluation_strategy` is deprecated and will be removed in version 4.46 of 🤗 Transformers. Use `eval_strategy` instead
  warnings.warn(
/content/modeleval.py:78: FutureWarning: `tokenizer` is deprecated and will be removed in version 5.0.0 for `Trainer.__init__`. Use `processing_class` instead.
  trainer = Trainer(


Fold 4 evaluation results: {'eval_f1': 0.0, 'eval_loss': 0.31710806488990784, 'eval_runtime': 11.7547, 'eval_samples_per_second': 142.496, 'eval_steps_per_second': 17.865, 'epoch': 3.0}
Training fold 5...


Some weights of RobertaForSequenceClassification were not initialized from the model checkpoint at roberta-base and are newly initialized: ['classifier.dense.bias', 'classifier.dense.weight', 'classifier.out_proj.bias', 'classifier.out_proj.weight']
You should probably TRAIN this model on a down-stream task to be able to use it for predictions and inference.
/usr/local/lib/python3.11/dist-packages/transformers/training_args.py:1575: FutureWarning: `evaluation_strategy` is deprecated and will be removed in version 4.46 of 🤗 Transformers. Use `eval_strategy` instead
  warnings.warn(
/content/modeleval.py:78: FutureWarning: `tokenizer` is deprecated and will be removed in version 5.0.0 for `Trainer.__init__`. Use `processing_class` instead.
  trainer = Trainer(


Fold 5 evaluation results: {'eval_f1': 0.41379310344827586, 'eval_loss': 0.2682809829711914, 'eval_runtime': 11.7165, 'eval_samples_per_second': 142.961, 'eval_steps_per_second': 17.923, 'epoch': 3.0}
Average evaluation results from cross-validation: {'eval_f1': 0.41010460372568264, 'eval_loss': 0.3129347741603851, 'eval_runtime': 11.772839999999999, 'eval_samples_per_second': 142.283, 'eval_steps_per_second': 17.8384, 'epoch': 3.0}


{'eval_f1': 0.41010460372568264,
 'eval_loss': 0.3129347741603851,
 'eval_runtime': 11.772839999999999,
 'eval_samples_per_second': 142.283,
 'eval_steps_per_second': 17.8384,
 'epoch': 3.0}

As a baseline, we find that simply fine-tuning RoBERTa on the data set yields an average F1 Score of $0.4101$. This is low, and is likely a result of the large imbalance between positive and negative samples -- there are roughly 10 times more negative samples than positive
samples.

## Simple Oversampling / Undersampling
In order to combat this label imbalance, we perform a naive oversampling / undersampling of the dataset. We resample the minority class five times, and undersample the majority class by a factor of two. This yields a balanced data set.

In [2]:
class NaiveResampling(AugmentationStrategy):

    def augment_data(self, data: Dataset) -> Dataset:
        print("Hello")